### ISSUE TITLE
BGPLVM error in expectation when passing in slice

### ISSUE THREAD

**Issue by alexisboukouvalas** [GPflow/GPflow#377]
Run into this when debugging a failure in the GPLVM notebook.
When passing in slice for active dimensions code fails with cholesky error.
When passing in active dimensions as lists, the code works. (change fHmmm to False)
@markvdw any idea on how to debug this further?
```
import GPflow
from GPflow import ekernels
from GPflow import kernels
import numpy as np
import pods
pods.datasets.overide_manual_authorize = True  # dont ask to authorize
np.random.seed(42)
GPflow.settings.numerics.quadrature = 'error'  # throw error if quadrature is used for kernel expectations
data = pods.datasets.oil_100()
Y = data['X']
Q = 5
M = 10  # number of inducing pts
N = Y.shape[0]
X_mean = GPflow.gplvm.PCA_reduce(Y, Q) # Initialise via PCA
Z = np.random.permutation(X_mean.copy())[:M]

fHmmm = True
if(fHmmm):
    k = ekernels.Add([ekernels.RBF(3, ARD=True, active_dims=slice(0,3)),
                  ekernels.Linear(2, ARD=False, active_dims=slice(3,5))])
else:
    k = ekernels.Add([ekernels.RBF(3, ARD=True, active_dims=[0,1,2]),
                  ekernels.Linear(2, ARD=False, active_dims=[3, 4])])
    
m = GPflow.gplvm.BayesianGPLVM(X_mean=X_mean, X_var=0.1*np.ones((N, Q)), Y=Y,
                                kern=k, M=M, Z=Z)
m.likelihood.variance = 0.01
m.optimize(disp=True, max_iters=100)
```

**Comment by markvdw** [comment]
Ok, the code is a bit complicated here, since there are quite a few cases. I suspect the error is in `Add.eKzxKxz()`.

Your kernels work on separate dimensions and the GPLVM is initialised as with Xcov being factorised over dimensions. In the case where `active_dims` is a list, we can determine that the kernel works on separate dimensions, and so ekernels.py:191 should be taken. This is simple and works as you mention.

When `active_dims` is a slice, things are more complicated. There is an analytic expression for the cross terms only for the case where _all dimensions are shared_. Currently we test for this by checking `active_dims` is a slice.  This is where the error is. We should be checking that it's a slice that covers all dimensions, and falls back to quadrature if not.

So there are two solutions:
- Check whether `active_dims` covers all dimensions for _both_ kernels in ekernels.py:221
- Modify the code in `Add.Linear_RBF_eKxzKzx()`. This shouldn't be too difficult.

The short term solution for you is to not use slices but lists :)

**Comment by st--** [comment]
This is actually still an issue - moreover, there seems to be a bug in the expectations code, even the `fHmmm = False` branch doesn't work! I've updated the MWE to work with gpflow2.

**Comment by ialong** [comment]
It's been a while but yes that's probably due to `active_dims` not being always supported for kernel expectations (#614). @st--  could you post the error the revised MWE produces?

**Comment by st--** [comment]
For the `fHmmm = False` case (analytic expectation), the traceback (without the tf.function, cut to the relevant part) is:
```
  File "/home/st/Code/gpflow2/gpflow/models/gplvm.py", line 148, in log_likelihood
    (self.kernel, self.inducing_variable)),
  File "/home/st/Code/gpflow2/gpflow/expectations/expectations.py", line 54, in expectation
    return dispatch.expectation(p, obj1, feat1, obj2, feat2, nghp=nghp)
  File "/home/st/anaconda3/envs/gpflow2/lib/python3.7/site-packages/multipledispatch/dispatcher.py", line 278, in __call__
    return func(*args, **kwargs)
  File "/home/st/Code/gpflow2/gpflow/expectations/sums.py", line 82, in _E
    crossexps.append(expectation(p, (k1, feat1), (k1, feat1), nghp=nghp))
  File "/home/st/Code/gpflow2/gpflow/expectations/expectations.py", line 54, in expectation
    return dispatch.expectation(p, obj1, feat1, obj2, feat2, nghp=nghp)
  File "/home/st/anaconda3/envs/gpflow2/lib/python3.7/site-packages/multipledispatch/dispatcher.py", line 278, in __call__
    return func(*args, **kwargs)
  File "/home/st/Code/gpflow2/gpflow/expectations/squared_exponentials.py", line 138, in _E
    if kern1.on_separate_dims(kern2) and isinstance(p, DiagonalGaussian):  # no joint expectations required
  File "/home/st/Code/gpflow2/gpflow/kernels/base.py", line 68, in on_separate_dims
    if self.active_dims is None or other.active_dims:
ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
```
For `fHmmm = True` (quadrature case), the code is super slow and memory hungry - presumably due to constructing 100^5 = 1e10 quadrature points! (Changing latent_dim to 2 and amending active_dims to be slice(0, 1) and slice(1, 2) respectively does seem to run fine.)

**Comment by ialong** [comment]
hmm the error message above can only occur if `other.active_dims` is a numpy array (which is not supported I guess). This is unrelated to the expectations code I think? It seems like it might be something to do with tf 1 -> tf 2

**Comment by ialong** [comment]
actually, upon further inspection, shouldn't line 68 in kernels/base.py be:
```python
if self.active_dims is None or other.active_dims is None:
```
instead of:
```python
if self.active_dims is None or other.active_dims:
```
?

**Comment by st--** [comment]
Good catch, I think that is definitely a bug. Fixed in #1178.
I guess the active_dims=slice(...) issue remains open though?

**Comment by ialong** [comment]
yes of course. I'll have a look once I update to gpflow2 (in the next few days). Though I think supporting `active_dims` is not a 1-minute fix for analytic expectations unfortunately. If somebody needs this and is willing to implement it, I'd be happy to provide assistance

**Comment by st--** [comment]
@ialong hi did you have a chance to look at this in GPflow 2 ? would be great if you could confirm whether or not this is still an issue in the gpflow 2.0 release. Thanks :)

**Comment by ialong** [comment]
Hi @st-- unfortunately I have not been using GPflow much these days (and have not upgraded to GPflow 2 yet in the end). If people need this I can have a look after the NeurIPS deadline. Maybe the best course of action for now is simply to raise an error message saying slices are not supported as `active_dims`, only lists?

**Comment by ialong** [comment]
kernel expectations are a bit of a rabbit hole so I want to make sure that if I put my hands on them I have enough time to do it properly. Very easy to introduce new bugs otherwise

**Comment by st--** [comment]
Makes sense! if you do want to look at it after NeurIPS that'd be great, and join us on the gpflow.slack.com chat (invite link in the README.md) if you want to stay in touch with the gpflow community :)
